File: tools/automation/src/asic-work-send-evidence.ts
import { createHash } from "node:crypto";
import { chmod, mkdir, readFile, rename, stat, unlink, writeFile } from "node:fs/promises";
import path from "node:path";

import {
  internalCommandSpec,
  type AsicInitializationEvidence,
  type AsicWorkSendEvidence,
  type AutomationCategory,
} from "./contracts.generated.js";
import type { ProcessPort } from "./process.js";
import { assertWithinWorkspace } from "./workspace.js";

export type AsicWorkSendEvidenceOptions = {
  readonly sourceProjection: string;
  readonly attemptSourceCommit: string;
  readonly projection: string;
};

type FailureCategory = Extract<AutomationCategory, "evidence_invalid" | "process_failed">;

const expectedSourceProjection =
  "docs/parity/evidence/asic002-initialization/asic-initialization-projection.json";

const corePaths = [
  "crates/bitaxe-asic/src/bm1366/work.rs",
  "crates/bitaxe-asic/src/bm1366/production.rs",
  "crates/bitaxe-asic/src/bm1366/command.rs",
] as const;

const spanContracts = [
  {
    path: "firmware/bitaxe/src/production_mining_session/asic_worker.rs",
    start: "                        AsicWorkerCommand::Dispatch {",
    end: "                        AsicWorkerCommand::Poll {",
  },
  {
    path: "firmware/bitaxe/src/asic_adapter/production.rs",
    start: "            Bm1366ProductionCommand::SendProductionWork(_) => {",
    end: "            Bm1366ProductionCommand::ReadProductionResult => {",
  },
  {
    path: "firmware/bitaxe/src/asic_adapter/production.rs",
    start: "        Bm1366AdapterAction::WriteFrame(frame) => {",
    end: "        Bm1366AdapterAction::ReadExact { len, timeout_ms } => {",
  },
] as const;

export class AsicWorkSendEvidenceError extends Error {
  public constructor(
    public readonly category: FailureCategory,
    message: string,
    public readonly publicValue: Readonly<Record<string, unknown>>,
  ) {
    super(message);
    this.name = "AsicWorkSendEvidenceError";
  }
}

function failure(category: FailureCategory, message: string): AsicWorkSendEvidenceError {
  return new AsicWorkSendEvidenceError(category, message, {
    stage: "sealed_work_send_projection",
    hardware_rerun_used: false,
    projection_published: false,
  });
}

function sha256(value: string | Buffer): string {
  return createHash("sha256").update(value).digest("hex");
}

function lowerHex(value: string, length: number): boolean {
  return value.length === length && new RegExp(`^[0-9a-f]{${String(length)}}$`, "u").test(value);
}

async function childText(
  processPort: ProcessPort,
  program: string,
  args: readonly string[],
  context: string,
): Promise<string> {
  try {
    const outcome = await processPort.run(internalCommandSpec(program, [...args], (value) => value));
    if (outcome.timedOut || outcome.exitCode !== 0) {
      throw failure("evidence_invalid", `${context} did not pass`);
    }
    return outcome.stdout.trim();
  } catch (error) {
    if (error instanceof AsicWorkSendEvidenceError) throw error;
    throw failure("process_failed", `${context} launch failed`);
  }
}

async function requireAbsent(candidate: string, context: string): Promise<void> {
  try {
    await stat(candidate);
    throw failure("evidence_invalid", `${context} must be absent before projection`);
  } catch (error) {
    if (error instanceof AsicWorkSendEvidenceError) throw error;
    if ((error as NodeJS.ErrnoException).code !== "ENOENT") throw error;
  }
}

function parseSource(document: string): AsicInitializationEvidence {
  let value: unknown;
  try {
    value = JSON.parse(document);
  } catch {
    throw failure("evidence_invalid", "source initialization projection is malformed");
  }
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw failure("evidence_invalid", "source initialization projection must be an object");
  }
  return value as AsicInitializationEvidence;
}

function validateSourceFacts(
  source: AsicInitializationEvidence,
  attemptSourceCommit: string,
): void {
  if (source.schema_version !== "bitaxe-asic-initialization-evidence-v1"
    || source.board !== 205
    || source.attempt_source_commit !== attemptSourceCommit
    || !source.package_admitted
    || source.runtime_identity !== "trusted"
    || source.runtime_attestation_status !== "trusted"
    || source.campaign_terminal_category !== "submit_response_observed"
    || source.submit_outcome !== "accepted"
    || source.safety_status !== "fresh"
    || !source.mine_on_boot_disabled
    || !source.safe_stop_confirmed
    || !source.lease_cleanup_confirmed
    || !source.usb_cleanup_ready
    || source.hardware_rerun_used
    || source.redaction_status !== "passed"
    || !source.initialization.mining_ready_initialization_completed
    || !source.initialization.production_uart_retained
    || !source.initialization.live_initialized_work_observed) {
    throw failure("evidence_invalid", "source initialization projection quorum is incomplete");
  }
}

function extractUniqueSpan(document: string, start: string, end: string): string {
  const startIndex = document.indexOf(start);
  if (startIndex === -1 || document.indexOf(start, startIndex + start.length) !== -1) {
    throw failure("evidence_invalid", "work-send source span start is not unique");
  }
  const endIndex = document.indexOf(end, startIndex + start.length);
  if (endIndex === -1 || document.indexOf(end, endIndex + end.length) !== -1) {
    throw failure("evidence_invalid", "work-send source span end is not unique");
  }
  return document.slice(startIndex, endIndex);
}

async function validateSourceCompatibility(
  processPort: ProcessPort,
  gitProgram: string,
  attemptSourceCommit: string,
  currentSourceCommit: string,
): Promise<void> {
  await childText(
    processPort,
    gitProgram,
    ["diff", "--quiet", attemptSourceCommit, currentSourceCommit, "--", ...corePaths],
    "work-send core source compatibility",
  );

  for (const span of spanContracts) {
    const [attemptDocument, currentDocument] = await Promise.all([
      childText(
        processPort,
        gitProgram,
        ["show", `${attemptSourceCommit}:${span.path}`],
        "attempt work-send source span",
      ),
      childText(
        processPort,
        gitProgram,
        ["show", `${currentSourceCommit}:${span.path}`],
        "current work-send source span",
      ),
    ]);
    if (extractUniqueSpan(attemptDocument, span.start, span.end)
      !== extractUniqueSpan(currentDocument, span.start, span.end)) {
      throw failure("evidence_invalid", "work-send source span drifted");
    }
  }
}

export async function projectAsicWorkSendEvidence(
  workspaceRoot: string,
  options: AsicWorkSendEvidenceOptions,
  processPort: ProcessPort,
  gitProgram: string,
  sourceValidatorProgram: string,
  validatorProgram: string,
): Promise<AsicWorkSendEvidence> {
  if (!lowerHex(options.attemptSourceCommit, 40)) {
    throw failure("evidence_invalid", "attempt source commit is invalid");
  }
  const sourceProjection = assertWithinWorkspace(workspaceRoot, options.sourceProjection);
  const projection = assertWithinWorkspace(workspaceRoot, options.projection);
  if (path.relative(workspaceRoot, sourceProjection) !== expectedSourceProjection) {
    throw failure("evidence_invalid", "source initialization projection path is invalid");
  }
  const candidate = `${projection}.candidate`;
  await requireAbsent(projection, "public projection");
  await requireAbsent(candidate, "projection candidate");

  const sourceDocument = await readFile(sourceProjection, "utf8");
  await childText(
    processPort,
    sourceValidatorProgram,
    [sourceProjection],
    "source initialization evidence validation",
  );
  const source = parseSource(sourceDocument);
  validateSourceFacts(source, options.attemptSourceCommit);

  const currentSourceCommit = await childText(
    processPort,
    gitProgram,
    ["rev-parse", "HEAD"],
    "current source identity",
  );
  const referenceCommit = await childText(
    processPort,
    gitProgram,
    ["-C", path.join(workspaceRoot, "reference/esp-miner"), "rev-parse", "HEAD"],
    "reference source identity",
  );
  if (!lowerHex(currentSourceCommit, 40)) {
    throw failure("evidence_invalid", "current work-send source identity is invalid");
  }
  if (!lowerHex(referenceCommit, 40)) {
    throw failure("evidence_invalid", "reference work-send source identity is invalid");
  }
  if (!lowerHex(source.current_source_commit, 40)) {
    throw failure("evidence_invalid", "source projection identity is invalid");
  }
  await childText(
    processPort,
    gitProgram,
    ["cat-file", "-e", `${options.attemptSourceCommit}^{commit}`],
    "attempt source admission",
  );
  await childText(
    processPort,
    gitProgram,
    ["merge-base", "--is-ancestor", source.current_source_commit, currentSourceCommit],
    "source projection ancestry",
  );
  await childText(
    processPort,
    gitProgram,
    ["ls-files", "--error-unmatch", expectedSourceProjection],
    "source projection tracking",
  );
  await validateSourceCompatibility(
    processPort,
    gitProgram,
    options.attemptSourceCommit,
    currentSourceCommit,
  );
  const relevantPaths = [
    expectedSourceProjection,
    ...corePaths,
    ...new Set(spanContracts.map((span) => span.path)),
  ];
  const worktreeState = await childText(
    processPort,
    gitProgram,
    ["status", "--porcelain", "--", ...relevantPaths],
    "work-send worktree compatibility",
  );
  if (worktreeState !== "") {
    throw failure("evidence_invalid", "work-send evidence paths have uncommitted drift");
  }

  const initializationProjectionSha256 = sha256(sourceDocument);
  const requestSha256 = sha256(JSON.stringify({
    command: "project-asic-work-send-evidence",
    source_projection: expectedSourceProjection,
    attempt_source_commit: options.attemptSourceCommit,
    current_source_commit: currentSourceCommit,
    projection: path.relative(workspaceRoot, projection),
  }));
  const evidence: AsicWorkSendEvidence = {
    schema_version: "bitaxe-asic-work-send-evidence-v1",
    board: 205,
    attempt_source_commit: options.attemptSourceCommit,
    current_source_commit: currentSourceCommit,
    reference_commit: referenceCommit,
    workflow: {
      schema_version: "bitaxe-workflow-identity-v1",
      command: "project-asic-work-send-evidence",
      request_sha256: requestSha256,
    },
    source: {
      initialization_projection_sha256: initializationProjectionSha256,
      initialization_projection_current_commit: source.current_source_commit,
      initialization_projection_valid: true,
    },
    work_send: {
      payload_length_bytes: 82,
      frame_length_bytes: 88,
      job_id_step: 8,
      job_id_modulus: 128,
      typed_write_frame_action: true,
      production_ready_gate_required: true,
      live_work_observed: true,
      qualified_result_observed: true,
      accepted_submit_observed: true,
      production_uart_retained: true,
      core_paths_unchanged: true,
      compatible_core_path_count: corePaths.length,
      dispatch_spans_unchanged: true,
      uart_write_span_unchanged: true,
    },
    package_admitted: true,
    runtime_identity: "trusted",
    runtime_attestation_status: "trusted",
    campaign_terminal_category: "submit_response_observed",
    submit_outcome: "accepted",
    safety_status: "fresh",
    mine_on_boot_disabled: true,
    safe_stop_confirmed: true,
    lease_cleanup_confirmed: true,
    usb_cleanup_ready: true,
    hardware_rerun_used: false,
    redaction_status: "passed",
  };

  await mkdir(path.dirname(projection), { recursive: true });
  try {
    await writeFile(candidate, `${JSON.stringify(evidence, null, 2)}\n`, {
      encoding: "utf8",
      flag: "wx",
      mode: 0o600,
    });
    await chmod(candidate, 0o600);
    await childText(processPort, validatorProgram, [candidate], "independent evidence validation");
    await rename(candidate, projection);
    await chmod(projection, 0o644);
  } catch (error) {
    await unlink(candidate).catch(() => undefined);
    throw error;
  }
  return evidence;
}

File: crates/bitaxe-automation-contracts/src/asic_work_send_evidence.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::{AutomationCommand, WorkflowIdentity, ASIC_WORK_SEND_EVIDENCE_SCHEMA};

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
pub struct AsicWorkSendSourceEvidence {
    pub initialization_projection_sha256: String,
    pub initialization_projection_current_commit: String,
    pub initialization_projection_valid: bool,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
pub struct AsicWorkSendObservationEvidence {
    pub payload_length_bytes: u64,
    pub frame_length_bytes: u64,
    pub job_id_step: u64,
    pub job_id_modulus: u64,
    pub typed_write_frame_action: bool,
    pub production_ready_gate_required: bool,
    pub live_work_observed: bool,
    pub qualified_result_observed: bool,
    pub accepted_submit_observed: bool,
    pub production_uart_retained: bool,
    pub core_paths_unchanged: bool,
    pub compatible_core_path_count: u64,
    pub dispatch_spans_unchanged: bool,
    pub uart_write_span_unchanged: bool,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
pub struct AsicWorkSendEvidence {
    pub schema_version: String,
    pub board: u16,
    pub attempt_source_commit: String,
    pub current_source_commit: String,
    pub reference_commit: String,
    pub workflow: WorkflowIdentity,
    pub source: AsicWorkSendSourceEvidence,
    pub work_send: AsicWorkSendObservationEvidence,
    pub package_admitted: bool,
    pub runtime_identity: String,
    pub runtime_attestation_status: String,
    pub campaign_terminal_category: String,
    pub submit_outcome: String,
    pub safety_status: String,
    pub mine_on_boot_disabled: bool,
    pub safe_stop_confirmed: bool,
    pub lease_cleanup_confirmed: bool,
    pub usb_cleanup_ready: bool,
    pub hardware_rerun_used: bool,
    pub redaction_status: String,
}

impl AsicWorkSendEvidence {
    pub fn validate(&self) -> Result<(), &'static str> {
        if self.schema_version != ASIC_WORK_SEND_EVIDENCE_SCHEMA || self.board != 205 {
            return Err("ASIC work-send evidence schema or board is invalid");
        }
        if self.workflow.schema_version != "bitaxe-workflow-identity-v1"
            || self.workflow.command != AutomationCommand::ProjectAsicWorkSendEvidence
        {
            return Err("ASIC work-send workflow identity is invalid");
        }
        for commit in [
            self.attempt_source_commit.as_str(),
            self.current_source_commit.as_str(),
            self.reference_commit.as_str(),
            self.source
                .initialization_projection_current_commit
                .as_str(),
        ] {
            if !is_lower_hex(commit, 40) {
                return Err("ASIC work-send source identity is invalid");
            }
        }
        for digest in [
            self.workflow.request_sha256.as_str(),
            self.source.initialization_projection_sha256.as_str(),
        ] {
            if !is_lower_hex(digest, 64) {
                return Err("ASIC work-send digest is invalid");
            }
        }
        if !self.source.initialization_projection_valid {
            return Err("ASIC work-send source evidence is invalid");
        }

        let work_send = &self.work_send;
        if work_send.payload_length_bytes != 82
            || work_send.frame_length_bytes != 88
            || work_send.job_id_step != 8
            || work_send.job_id_modulus != 128
            || !work_send.typed_write_frame_action
            || !work_send.production_ready_gate_required
            || !work_send.live_work_observed
            || !work_send.qualified_result_observed
            || !work_send.accepted_submit_observed
            || !work_send.production_uart_retained
            || !work_send.core_paths_unchanged
            || work_send.compatible_core_path_count != 3
            || !work_send.dispatch_spans_unchanged
            || !work_send.uart_write_span_unchanged
        {
            return Err("ASIC work-send observation is incomplete");
        }
        if !self.package_admitted
            || self.runtime_identity != "trusted"
            || self.runtime_attestation_status != "trusted"
            || self.campaign_terminal_category != "submit_response_observed"
            || self.submit_outcome != "accepted"
            || self.safety_status != "fresh"
            || !self.mine_on_boot_disabled
            || !self.safe_stop_confirmed
            || !self.lease_cleanup_confirmed
            || !self.usb_cleanup_ready
            || self.hardware_rerun_used
            || self.redaction_status != "passed"
        {
            return Err("ASIC work-send campaign or cleanup evidence is invalid");
        }
        Ok(())
    }
}

fn is_lower_hex(value: &str, length: usize) -> bool {
    value.len() == length
        && value
            .bytes()
            .all(|byte| byte.is_ascii_hexdigit() && !byte.is_ascii_uppercase())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn evidence() -> AsicWorkSendEvidence {
        AsicWorkSendEvidence {
            schema_version: ASIC_WORK_SEND_EVIDENCE_SCHEMA.to_owned(),
            board: 205,
            attempt_source_commit: "a".repeat(40),
            current_source_commit: "b".repeat(40),
            reference_commit: "c".repeat(40),
            workflow: WorkflowIdentity {
                schema_version: "bitaxe-workflow-identity-v1".to_owned(),
                command: AutomationCommand::ProjectAsicWorkSendEvidence,
                request_sha256: "d".repeat(64),
            },
            source: AsicWorkSendSourceEvidence {
                initialization_projection_sha256: "e".repeat(64),
                initialization_projection_current_commit: "f".repeat(40),
                initialization_projection_valid: true,
            },
            work_send: AsicWorkSendObservationEvidence {
                payload_length_bytes: 82,
                frame_length_bytes: 88,
                job_id_step: 8,
                job_id_modulus: 128,
                typed_write_frame_action: true,
                production_ready_gate_required: true,
                live_work_observed: true,
                qualified_result_observed: true,
                accepted_submit_observed: true,
                production_uart_retained: true,
                core_paths_unchanged: true,
                compatible_core_path_count: 3,
                dispatch_spans_unchanged: true,
                uart_write_span_unchanged: true,
            },
            package_admitted: true,
            runtime_identity: "trusted".to_owned(),
            runtime_attestation_status: "trusted".to_owned(),
            campaign_terminal_category: "submit_response_observed".to_owned(),
            submit_outcome: "accepted".to_owned(),
            safety_status: "fresh".to_owned(),
            mine_on_boot_disabled: true,
            safe_stop_confirmed: true,
            lease_cleanup_confirmed: true,
            usb_cleanup_ready: true,
            hardware_rerun_used: false,
            redaction_status: "passed".to_owned(),
        }
    }

    #[test]
    fn complete_closed_projection_is_accepted() {
        // Arrange
        let candidate = evidence();

        // Act
        let result = candidate.validate();

        // Assert
        assert_eq!(result, Ok(()));
    }

    #[test]
    fn incompatible_dispatch_span_is_rejected() {
        // Arrange
        let mut candidate = evidence();
        candidate.work_send.dispatch_spans_unchanged = false;

        // Act
        let result = candidate.validate();

        // Assert
        assert_eq!(result, Err("ASIC work-send observation is incomplete"));
    }

    #[test]
    fn nonaccepted_submit_is_rejected() {
        // Arrange
        let mut candidate = evidence();
        candidate.submit_outcome = "rejected".to_owned();

        // Act
        let result = candidate.validate();

        // Assert
        assert_eq!(
            result,
            Err("ASIC work-send campaign or cleanup evidence is invalid")
        );
    }
}
